sql/src/main/java/io/crate/operation/projectors/sharding/ShardingUpsertExecutor.java

```diff
@@ -229,6 +232,7 @@ public void onFailure(Exception e) {
 
         private void countdown() {
             if (numRequests.decrementAndGet() == 0) {
+                Exception interruptedException = interrupt.get();
```

Comment:
shouldn't the interrupt be handled regardless of the remaining outstanding requests? even more, the entire operation should be failed (I think even with this fix, killing a copy from task will not  actually kill it)

Reply:
If this would already set the future we'd just "fake" a interrupt/faster response, but the other operations would still be ongoing.

Generally this shouldn't be much of an issue here anyway, because if a kill is invoked it's registered on each node, and the BatchIterators should start failing anyway.

Reply:
But we can stop consuming from the source batch iterator when the exception is encountered. I don't think there's any point in continuing and not shutting the operation down as soon as the exception is encountered. 

When looking in the "kill copy task" impediment I'm pretty sure the source batch iterator did not receive the kill signal.

Reply:
> But we can stop consuming from the source batch iterator when the exception is encountered

That's not what's happening in either case. The `BatchIteratorBackpressureExecutor` doesn't stop earlier if the execution fails. It only stops if `BatchIterator.moveNext` fails - which is the case if a kill happens.

The change in this PR mainly affects the case when the source is already consumed and only the "last" requests are being processed - which are then interrupted.

> I don't think there's any point in continuing and not shutting the operation down as soon as the exception is encountered.

The point is that we *cannot* shutdown the remote operations that are already in progress. There are two choices: Wait or ignore and return early. Ignore and return early has the problem that it's a) confusing for a user if there is still load, although he got a response on the kill;  b) It can lead to problems when we internally kill as the resources are still being accessed. That's also confusing if you're debugging things.

Reply:
ok, numRequests is per bulk not per operation so we'll wait for the in-progress bulk to finish (I was under the impression it's the entire operation)